man/fi_mr.3.md

```diff
@@ -736,6 +743,10 @@ The follow flag may be specified to any memory registration call.
   API AllocHost call (as opposed to using malloc-like host allocation,
   unified/shared memory allocation, or AllocDevice).
 
+*FI_HMEM_DATA_GDRCOPY_HANDLE*
+: This flag indicates that hmem_data points to a gdrcopy handle. This flag is
+  ignored unless iface is set to FI_HMEM_CUDA.
```

Comment:
If this is defined and exposed to users, it needs to be well defined, so that it's usable.  If we're only going to use this for some internal calls, then we can use one of the reserved internal flags, with some internal structure.

These are the relevant gdrcopy calls:
```
gdr_pin_buffer(gdr_t g, unsigned long addr, size_t size, uint64_t p2p_token, uint32_t va_space, gdr_mh_t *handle)
gdr_map(gdr_t g, gdr_mh_t handle, void **va, size_t size);
gdr_copy_to_mapping(gdr_mh_t handle, void *map_d_ptr, const void *h_ptr, size_t size)
```
This is the flow I anticipated:
```
gdr_mh_t gdr_mem_handle;
void *gdr_mapped_addr;
struct fi_mr_attr mr_attr;
struct iovec iov;

gdr_pin_buffer(g, addr, len, ?, ?, &gdr_mem_handle);
gdr_map(g, gdr_mem_handle, &gdr_mapped_addr, len);

/* Register the gdr mapped buffer */
iov.iov_base = gdr_mapped_addr;
iov.iov_len = len;
mr_attr.iov = &iov;
mr_attr. ...
mr_attr.hmem_data = gdr_mem_handle;
fi_mr_regattr(..., &mr_attr, FI_HMEM_GDR_MAPPED);
```

The grpcopy routine needs the gdr_mem_handle and gdr_mapped_addr.

Reply:
This makes sense if gdr pinning and address translation are delegated to the application, e.g. application ensures page alignment, and always passes page-aligned addresses to libfabric.

It's different in the current gdrcopy implementation. libfabric encapsulates this process from the application, and caches the alignment information in `struct gdrcopy_handle`, used for address translation during `gdr_copy_from/to_mapping`.

>  If we're only going to use this for some internal calls, then we can use one of the reserved internal flags, with some internal structure.

This is a good suggestion. In this case an internal flag & struct should be clear and sufficient. I will propose a change accordingly.